### ISSUE TITLE
[clippy] arc_with_non_send_sync - interrupt_lock

### ISSUE THREAD

**Issue by gwenn** [rusqlite/rusqlite#1358]
```
error: usage of `Arc<T>` where `T` is not `Send` or `Sync`
  --> src/inner_connection.rs:48:29
   |
48 |             interrupt_lock: Arc::new(Mutex::new(db)),
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider using `Rc<T>` instead or wrapping `T` in a std::sync type like `Mutex<T>`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arc_with_non_send_sync
   = note: `#[deny(clippy::arc_with_non_send_sync)]` on by default
```
Clippy seems to suggest that the `Arc` / `Mutex` is useless because the connection cannot be shared by multiple threads, no ?

**Comment by gwenn** [comment]
https://github.com/rust-lang/rust-clippy/issues/11076 => false positive

**Comment by nathaniel-daniel** [comment]
I think this is actually https://github.com/rust-lang/rust-clippy/issues/11382, as https://github.com/rust-lang/rust-clippy/issues/11076 has been closed already. There is no generic code here. Clippy seems confused as to why a pointer, which should make the `Arc` and the type that contains it not `Send`, is being used over an `Rc`, missing the [unsafe `Send` impl for `InnerConnection`](https://github.com/rusqlite/rusqlite/blob/36264dc3cc4bce89b9f20dc030d49d4c6a0045b8/src/inner_connection.rs#L40).

**Comment by gwenn** [comment]
https://github.com/rusqlite/rusqlite/pull/408/files
`*mut ffi::sqlite3` / `InterruptHandle` are `Send`able
So for me, Clippy is wrong (and PR #1397 is invalid).
But I am not completely sure.

**Comment by nathaniel-daniel** [comment]
> *mut ffi::sqlite3 / InterruptHandle are Sendable

Well technically, `mut ffi::sqlite3` is not (since its a pointer), while `InterruptHandle` is (since it is done with `unsafe impl Send`), which is tripping up Clippy. Clippy doesn't seem to consider `unsafe impl Send` stuff in this lint.

> So for me, Clippy is wrong (and PR https://github.com/rusqlite/rusqlite/pull/1397 is invalid).

Yeah, that PR looks unsafe; you can't share an `Rc` between multiple threads like that. I think I agree that it looks like a Clippy bug; this is an odd use case, but it seems valid and `Arc` definitely should be used here. This bug seems to be tracked by the issue I linked above. Maybe its worth adding an allow for the lint for that one line and linking the Clippy bug?

**Comment by gwenn** [comment]
> Maybe its worth adding an allow for the lint for that one line

Please do or I will.
But for me, it's not a Clippy bug because only `InterruptHandle` is `Send`able.
Thanks.

**Comment by gwenn** [comment]
Fixed with `rusqlite` version 0.30.0
